python_onnxifi/python_onnxifi.cc

```diff
@@ -0,0 +1,355 @@
+        if (onnxReleaseBackendID(ids_[i]) != ONNXIFI_STATUS_SUCCESS) {
+          throw std::runtime_error(
+              "Internal error: onnxReleaseBackendID failed (expected "
+              "ONNXIFI_STATUS_SUCCESS)");
+        }
+      }
+      delete[] ids_;
+    }
+  }
+
+  // Convenience maps to map betweeen onnxEnum and string
+  //  python user will input the string
+  //  C++ internals and ONNXIFI represent as onnxEnum
+  static const std::unordered_map<onnxEnum, std::string>&
+  deviceType_to_string_();
+  static const std::unordered_map<std::string, onnxEnum>&
+  string_to_deviceType_();
+
+ private:
+  // Array of all backendIDs
+  onnxBackendID* ids_;
+  size_t num_backends_;
+
+  // device_to_onnxBackendID[deviceType][deviceID - 1] will give corresponding
+  // onnxBackendID
+  // 1-indexed to conform to onnx/onnx/backend/base.py
+  // deviceID of 0 refers to any onnxBackendID of that deviceType
+  std::unordered_map<onnxEnum, std::vector<onnxBackendID>>
+      device_to_onnxBackendID_;
+
+  // If the backendID has been initialized, will map to corresponding backendID
+  // pointer
+  // Otherwise, not present
+  std::unordered_map<onnxBackendID, onnxBackend> initialized_;
+};
+
+const std::unordered_map<onnxEnum, std::string>&
+DeviceIDs::deviceType_to_string_() {
+  static std::unordered_map<onnxEnum, std::string> deviceType_to_string_ = {
+      {ONNXIFI_DEVICE_TYPE_NPU, "NPU"},
+      {ONNXIFI_DEVICE_TYPE_DSP, "DSP"},
+      {ONNXIFI_DEVICE_TYPE_GPU, "GPU"},
+      {ONNXIFI_DEVICE_TYPE_CPU, "CPU"},
+      {ONNXIFI_DEVICE_TYPE_FPGA, "FPGA"},
+      {ONNXIFI_DEVICE_TYPE_HETEROGENEOUS, "HETEROGENEOUS"}};
+  return deviceType_to_string_;
+}
+
+const std::unordered_map<std::string, onnxEnum>&
+DeviceIDs::string_to_deviceType_() {
+  static std::unordered_map<std::string, onnxEnum> string_to_deviceType_ = {
+      {"NPU", ONNXIFI_DEVICE_TYPE_NPU},
+      {"DSP", ONNXIFI_DEVICE_TYPE_DSP},
+      {"GPU", ONNXIFI_DEVICE_TYPE_GPU},
+      {"CPU", ONNXIFI_DEVICE_TYPE_CPU},
+      {"FPGA", ONNXIFI_DEVICE_TYPE_FPGA},
+      {"HETEROGENEOUS", ONNXIFI_DEVICE_TYPE_HETEROGENEOUS}};
+
+  return string_to_deviceType_;
+}
+
+struct BackendRep {
+ public:
+  //ONNXIFI graph that has been initalized
+  //takes ownership of serializedModel used to initialize G
+  BackendRep(std::string&& serializedModel, onnxBackend backend) : graph_(), backend_(backend), serialized_model_(serializedModel), conversion_() {
+    if (onnxInitGraph(backend, serialized_model_.size(), serialized_model_.c_str(), 0, nullptr, &graph_)
+                                         != ONNXIFI_STATUS_SUCCESS) {
+      std::cerr << "Could not initialize graph on given device" << std::endl;
+      graph_ = NULL; //Make dummy BackendRep with NULL graph TODO: Return python None object
+    }
+  }
+
+  // TODO: Implement run
+  //TODO: Change usage of memoryFence (onnxEvent instead of onnxEvent*) when submodule pulled
+  py::dict run(py::dict inputs, py::kwargs kwargs) {
```

Comment:
inputs should be const

Reply:
I tried this initially, but pybind11 array request function isn't declared const